pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -222,136 +306,220 @@ func (m *rangefeedMuxer) startNodeMuxRangeFeed(
-			if v, found := m.producers.Load(e.StreamID); found {
-				producer = (*channelRangeFeedEventProducer)(v)
+		if err := stuckWatcher.do(func() (err error) {
+			event, err = conn.receiver.Recv()
+			return err
+		}); err != nil {
+			return err
+		}
+
+		active := conn.lookupStream(event.StreamID)
+
+		// The stream may already have terminated. That's fine -- we may have
+		// encountered range split or similar rangefeed error, causing the caller to
+		// exit (and terminate this stream), but the server side stream termination
+		// is async and probabilistic (rangefeed registration output loop may have a
+		// checkpoint event available, *and* it may have context cancellation, but
+		// which one executes is a coin flip) and so it is possible that we may see
+		// additional event(s) arriving for a stream that is no longer active.
+		if active == nil {
+			if log.V(1) {
+				log.Infof(ctx, "received stray event stream %d: %v", event.StreamID, event)
 			}
+			continue
+		}
 
-			// The stream may already have terminated (either producer is nil, or
-			// producer.muxClientCtx.Done()). That's fine -- we may have encountered range
-			// split or similar rangefeed error, causing the caller to exit (and
-			// terminate this stream), but the server side stream termination is async
-			// and probabilistic (rangefeed registration output loop may have a
-			// checkpoint event available, *and* it may have context cancellation, but
-			// which one executes is a coin flip) and so it is possible that we may
-			// see additional event(s) arriving for a stream that is no longer active.
-			if producer == nil {
-				if log.V(1) {
-					log.Infof(ctx, "received stray event stream %d: %v", e.StreamID, e)
+		switch t := event.GetValue().(type) {
+		case *kvpb.RangeFeedCheckpoint:
+			if t.Span.Contains(active.Span) {
+				// If we see the first non-empty checkpoint, we know we're done with the catchup scan.
+				if !t.ResolvedTS.IsEmpty() && active.catchupRes != nil {
+					active.catchupRes.Release()
+					active.catchupRes = nil
 				}
-				continue
+				// Note that this timestamp means that all rows in the span with
+				// writes at or before the timestamp have now been seen. The
+				// Timestamp field in the request is exclusive, meaning if we send
+				// the request with exactly the ResolveTS, we'll see only rows after
+				// that timestamp.
+				active.startAfter.Forward(t.ResolvedTS)
 			}
+		case *kvpb.RangeFeedSSTable:
+		case *kvpb.RangeFeedError:
+			log.VErrEventf(ctx, 2, "RangeFeedError: %s", t.Error.GoError())
+			if active.catchupRes != nil {
+				m.ds.metrics.RangefeedErrorCatchup.Inc(1)
+			}
+			conn.streams.Delete(event.StreamID)
+			if _, err := handleRangefeedError(ctx, t.Error.GoError()); err != nil {
+				// If this is an error we cannot recover from, terminate the rangefeed.
+				return err
+			}
+			if err := m.restartActiveRangeFeed(ctx, active, t.Error.GoError()); err != nil {
+				return err
+			}
+			continue
```

Comment:
Can you write up the overall lifecycle somewhere? It isn't immediately clear to me why it's safe to reuse `conn` and `conn.receiver` here.